compatibility with --trace-gen-full

nakengelhardt opened the pull request:
The fifth and last in a series of PRs trying to cut #4 into somewhat more manageable chunks. [Click here](https://github.com/nakengelhardt/btor2tools/compare/fixups...nakengelhardt:full_trace) to view only the changes introduced in this PR (after #9).

This PR makes btorsim more tolerant of redundant information in the witness, to be able to use witnesses generated with `btormc --trace-gen-full`. It adds support for the operator `init <array> <vector>` that initializes all memory locations (which I was not previously aware of) and it adds support for parsing this initialization information from the witness as well - both in the format `<id> [*] <vector>` emitted by `btormc --trace-gen-full` or the format `<id> <vector>` emitted by [pono (cosa2)](https://github.com/upscale-project/pono).

mpreiner commented on src/btorsim/btorsim.cpp:
```diff
@@ -1137,6 +1165,11 @@ parse_assignment ()
     index_columno = columno + 1;
     while ((ch = next_char ()) == '0' || ch == '1')
       BTOR2_PUSH_STACK (array_index, ch);
+    if (ch == '*')
+    {
+      BTOR2_PUSH_STACK (array_index, ch);
+      ch = next_char();
+    }
```
Add a comment that this is btormc specific output. The BTOR2 witness format does not specify how to print initialized arrays, which is usually not required since the initialization information is in the BTOR2 model.

mpreiner commented on src/btorsim/btorsim.cpp:
```diff
@@ -1198,7 +1236,18 @@ parse_state_part (int64_t k)
+    }
     else
+    {
+      if BTOR2_EMPTY_STACK (array_index)
+      {
+        BTOR2_PUSH_STACK (array_index, '*');
+        BTOR2_PUSH_STACK (array_index, 0);
+      }
```
This is the pono/cosa2 specific output, right? Please add a comment.

nakengelhardt commented on src/btorsim/btorsimam.cpp:
```diff
@@ -0,0 +1,219 @@
+	return res;
+}
+
+bool BtorSimArrayModel::operator==(const BtorSimArrayModel& other) const
```
Before I add changes to fix issues with previous PRs, while looking it over I noticed that I had failed to consider a bunch of edge cases in array comparison, so this is an improvement to #10. This comparison function is now quite complex, please look it over carefully. I'm not sure yosys will ever emit a direct array comparison, so it won't be as well-tested by our users as the other changes...

nakengelhardt commented:
I believe that aside from [the question of where to put the documentation of the info file format](https://github.com/Boolector/btor2tools/pull/8#discussion_r464947140), I've now addressed all your feedback.

mpreiner reviewed (approved):
Thanks a lot for the changes!

nakengelhardt commented:
Sorry for taking so long for a minor change! I got distracted by being fired the day before you approved this PR, but now that that's all sorted out, here's the updated code.

mpreiner commented:
No worries and thanks for all your work!
